docs/specifications/linking-and-embedding-specification.md

```diff
@@ -0,0 +1,364 @@
+
+### Notes
+
+Notes in Foam are individual markdown files. Notes can both reside in the Foam workspace root and any of its sub-folders. Both files and folders are encouraged to follow the kebab-case naming convention (e.g. `some/path-to/a-file.md`).
+
+### Headings
+
+Headings in Foam are denoted with markdown heading syntax, e.g. `## Heading on Second Level`. Each heading will have a corresponding id, a kebab-case version of the text (e.g. `heading-on-second-level`). In case there's multiple headings with the same text, their id is suffixed with a running number (e.g. `my-heading`, `my-heading-1`, `my-heading-2`).
+
+### Anchors
+
+User can add arbitrary anchors into a note with HTML-syntax: `<a name="my-anchor"></a>`. Since these names (identifiers) are written by users, they might end up being non-unique within a note. In these cases, an error is shown to the user. See the reasoning of using the name-attribute instead of the id-attribute from https://stackoverflow.com/a/7335259. In the future, better syntax for the anchors might be developed.
+
+### EBlocks
+
+While anchors define a single location in the document, eblocks define two; the beginning and the end of an eblock. The intent of a block is to specify content bounded by the begin and end markers. The end of an eblock can either be explicit or implicit.
+
+**Explicit eblocks** are defined with the anchor syntax and have two special rules; 1) the name of the beginning anchor starts with `begin:` and the ending anchor starts with `end:`, and 2) the other parts of the name must match. An example of matching anchors is:
+ 
+  ```markdown
+  <a name="begin:my-block"></a>
+  Sunt cillum occaecat adipisicing aliquip excepteur
+  nisi ea nostrud do nostrud.
+
+  Ad magna labore tempor dolor dolore nulla magna Lorem
+  magna qui et. Ipsum eiusmod minim ad est nostrud sunt occaecat.
+  <a name="end:my-block"></a>
+  ```
+
+**Implicit eblocks** are defined by any anchor whose name is not starting with `begin:` or `end:`, and the point where the nesting level of the first non-immediate AST node either remains on the same level or decreases in comparison to the starting place of the anchor. The nesting level is changed e.g. with different levels of headings and bullet list indents. Additionally, the eblock can also be started with a heading, in which case the eblock ends when the chapter following the heading ends. In case the end of a file is found before the nesting level of the markdown content decreases, eblock ends there. The example below illustrate implicit eblocks with various eblock definitions.
+
+```markdown
+# My Note
+
+Occaecat esse minim Lorem fugiat mollit eu eiusmod magna dolor.
+Velit ea cupidatat pariatur amet pariatur est est laboris culpa
+ipsum magna eu nulla esse.
+
+## Second Level
+
+<a name="y"></a>
+Cupidatat velit esse est anim consequat enim ad.
+
+- Nostrud incididunt
+- <a name="x"></a> consequat non dolore
```

Comment:
What do you think about the ability to start an EBlock at the end of a text, as well as the start e.g.
```md
- consequat non dolore <a name="x" />
  - consequat fugiat
```

This would improve the readability of the list, since as a reader I can scan through the left side of the list and mentally stop reading when I hit the block marker.

Semantically, this would mean that a block cannot be started mid-paragraph, and would always include the entire parent AST node. This might be confusing.  Unsure. Leaving comment here anyway.

Reply:
From [a discussion on Discord](https://discordapp.com/channels/729975036148056075/730056943997288540/735539642979450990):

I'm not sure I agree with the example of implicit blocks for `x`. IMO `x` should refer only to `consequat non dolore`.

If you want it to refer to the entire bullet point, then I think it should be:
```
- Nostrud incididunt
<a name="x"></a>
- consequat non dolore
  - consequat fugiat
  - eiusmod esse ipsum
    - anim nulla aute
- ad reprehenderit
```

Otherwise the sibling of `x` is the content of the bullet point, not the bullet point itself.

I’m thinking in terms of AST: the anchor refers to its next sibling (that’s how I understood your spec).